### ISSUE TITLE
I think `optimizer.zero_grad()` should go before `loss.backward()`

### ISSUE THREAD

**Issue by geopapa11** [pytorch/tutorials#2799]
I think `optimizer.zero_grad()` should go before `loss.backward()` to ensure that gradients are zeroed out before the calculation of the gradients that takes place inside  `loss.backward()`.

https://github.com/pytorch/tutorials/blob/30e14df26091ef3ac58c8327e1e5c5cbf578f382/beginner_source/basics/quickstart_tutorial.py#L157

**Comment by jmarintur** [comment]
Hi @geopapa11, indeed, `optimizer.zero_grad()` should go before `loss.backward()`. I've just created a branch to modify it in my fork. Please let me know if you want to be the one creating the PR instead.

**Comment by geopapa11** [comment]
Thanks @jmarintur, I asked the person who made the corresponding change in this commit about a year ago to make sure I am not missing anything.
https://github.com/geopapa11/tutorials/commit/be6e86342233bdc932deace52d7388073fbb66b4

**Comment by jmarintur** [comment]
Hi @geopapa11 , I quickly jumped into it. I thought about possible scenarios this won't be ideal. For a simple training loop, as it is, should work too.

**Comment by jmarintur** [comment]
I was looking at more issues in this repo and I found that this was already mentioned in [issue#2693](https://github.com/pytorch/tutorials/issues/2693). In the comments, they pointed out that the accepted answer is to run first the  `optimizer.zero_grad() `, then `loss.backward()`, and finally `optimizer.step()`.

@svekars, for consistency, shouldn't we change it to the aforementioned order?

**Comment by albanD** [comment]
As discussed in https://github.com/pytorch/tutorials/pull/2800 I think we do want to keep the current order to ensure minimal memory use which is a generally important concern these days.

Thanks for bringing this up though!
